File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        int n = s.length();
        int m = p.length();
        
        int[] f1 = new int[26];
        int[] f2 = new int[26];
        List<Integer> ans = new ArrayList<>();

        for(int i=0;i<m;i++){
            f2[p.charAt(i) - 'a']++;
        }
        int i=0,j=0;
        while(j < n){
            f1[s.charAt(j) - 'a']++;
            if(j-i+1 >m){
                f1[s.charAt(i) -'a']--;
                i++;
            }
            if(Arrays.equals(f1,f2)){
                ans.add(i);
            }
            j++;
        }
        return ans;
        
    }
}

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int n = s.size();
        int m = p.size();

        vector<int> f1(26 ,0);
        vector<int> f2(26, 0);
        vector<int> ans;

        for(int i=0;i<m;i++){
            f2[p[i] - 'a']++;
        }
        int i=0,j=0;
        while(j < n){
            f1[s[j] - 'a']++;
            if(j-i+1 > m){
                f1[s[i] - 'a']--;
                i++;
            }
            if(f1 == f2){
                ans.push_back(i);
            }
            j++;
        }
        return ans;
    }
};